src/batcher/solve_pair.rs

```diff
@@ -1,16 +1,438 @@
 use crate::models::order::Order;
 use crate::models::solution::Solution;
 use anyhow::Result;
-use web3::types::U256;
+use web3::types::{U256, H256, Address};
+
+
+pub enum OrderPairType {
+    LhsFullyFilled,
+    RhsFullyFilled,
+    BothFullyFilled,
+}
+
+
+trait Matchable {
+    /// Returns which of the two orders can be fully matched, or both, or none.
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType>;
+    /// Returns whether the orders can be matched. For this the tokens need
+    /// to match, there must be a price that satisfies both orders.
+    fn attracts(&self, other: &Order) -> bool;
+    /// Returns whether this order's sell token is the other order's buy token
+    /// and vice versa.
+    fn opposite_tokens(&self, other: &Order) -> bool;
+    /// Returns whether there is a price that satisfies both orders.
+    fn have_price_overlap(&self, other: &Order) -> bool;
+}
+
+
+impl Matchable for Order {
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType> {
+        if !self.attracts(other) {
+            return None;
+        }
+
+        if self.buy_amount <= other.sell_amount && self.sell_amount <= other.buy_amount {
+            Some(OrderPairType::LhsFullyFilled)
+        } else if self.buy_amount >= other.sell_amount && self.sell_amount >= other.buy_amount {
+            Some(OrderPairType::RhsFullyFilled)
+        } else {
+            Some(OrderPairType::BothFullyFilled)
+        }
+    }
+
+    fn attracts(&self, other: &Order) -> bool {
+        self.opposite_tokens(other) && self.have_price_overlap(other)
+    }
+
+    fn opposite_tokens(&self, other: &Order) -> bool {
+        self.buy_token == other.sell_token && self.sell_token == other.buy_token
+    }
+
+    fn have_price_overlap(&self, other: &Order) -> bool {
+        self.sell_amount > U256::zero()
+            && other.sell_amount > U256::zero()
+            && self.buy_amount * other.buy_amount <= self.sell_amount * other.sell_amount
+    }
+}
+
+struct Match {
+    order_pair_type: OrderPairType,
+    orders: OrderPair,
+}
+
+type OrderPair = [Order; 2];
+
+type MatchedAmounts = [U256; 2];
+
 
 pub fn solve_pair(
     sell_orders_token0: &Vec<Order>,
     sell_orders_token1: &Vec<Order>,
 ) -> Result<Solution> {
+    // Can we assume that the orders are sorted already?
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token0));
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token1));
+    // Or do we need to sort them first here?
+    //sell_orders_token0.sort();
+    //sell_orders_token1.sort();
```

Comment:
Was this here before? I don't think we need this commented code if its not gonna be used... or was this a question?

Reply:
This was meant as a question since I don't know about the surroundings :shrug:

Reply:
It's gone.